File: crates/proof-napi/src/app_state.rs
use std::collections::BTreeMap;
use std::fs;
use std::path::{Path, PathBuf};
use std::time::Duration;

use proof_core::{VerificationLimits, inspect_package, load_workspace};
use rusqlite::{Connection, params};
use serde::Serialize;

#[cfg(test)]
use rusqlite::OptionalExtension;

use crate::BridgeFailure;

const SCHEMA_VERSION: i64 = 1;
const PREFERENCE_KEYS: &[&str] = &["language", "theme", "lastSection", "windowState"];

#[derive(Debug, Clone, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct RecentItem {
    pub path: String,
    pub last_opened_at: String,
}

#[derive(Debug, Clone, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct WorkbenchState {
    pub schema_version: i64,
    pub preferences: BTreeMap<String, String>,
    pub recent_workspaces: Vec<RecentItem>,
    pub recent_packages: Vec<RecentItem>,
}

pub fn initialize(database: &Path) -> Result<WorkbenchState, BridgeFailure> {
    let connection = open(database)?;
    migrate(&connection)?;
    read_state(&connection)
}

pub fn set_preference(
    database: &Path,
    key: &str,
    value: &str,
) -> Result<WorkbenchState, BridgeFailure> {
    if !PREFERENCE_KEYS.contains(&key) {
        return Err(BridgeFailure::input(
            "PREFERENCE_KEY_NOT_ALLOWED",
            "Preference key is not allowlisted.",
        ));
    }
    if value.len() > 2_000 {
        return Err(BridgeFailure::input(
            "PREFERENCE_VALUE_TOO_LONG",
            "Preference value exceeds 2000 bytes.",
        ));
    }
    let connection = open(database)?;
    migrate(&connection)?;
    connection
        .execute(
            "INSERT INTO preferences(key, value) VALUES(?1, ?2) \
             ON CONFLICT(key) DO UPDATE SET value = excluded.value",
            params![key, value],
        )
        .map_err(BridgeFailure::database)?;
    read_state(&connection)
}

pub fn remember_recent(
    database: &Path,
    kind: &str,
    path: &Path,
    timestamp: &str,
) -> Result<WorkbenchState, BridgeFailure> {
    validate_kind(kind)?;
    let connection = open(database)?;
    migrate(&connection)?;
    connection
        .execute(
            "INSERT INTO recent_items(kind, path, last_opened_at) VALUES(?1, ?2, ?3) \
             ON CONFLICT(kind, path) DO UPDATE SET last_opened_at = excluded.last_opened_at",
            params![kind, path.to_string_lossy(), timestamp],
        )
        .map_err(BridgeFailure::database)?;
    connection
        .execute(
            "DELETE FROM recent_items WHERE rowid IN (\
               SELECT rowid FROM recent_items WHERE kind = ?1 \
               ORDER BY last_opened_at DESC, rowid DESC LIMIT -1 OFFSET 20\
             )",
            params![kind],
        )
        .map_err(BridgeFailure::database)?;
    read_state(&connection)
}

pub fn rebuild_recents(database: &Path) -> Result<WorkbenchState, BridgeFailure> {
    let connection = open(database)?;
    migrate(&connection)?;
    let mut statement = connection
        .prepare("SELECT kind, path FROM recent_items")
        .map_err(BridgeFailure::database)?;
    let candidates = statement
        .query_map([], |row| {
            Ok((row.get::<_, String>(0)?, row.get::<_, String>(1)?))
        })
        .map_err(BridgeFailure::database)?
        .collect::<Result<Vec<_>, _>>()
        .map_err(BridgeFailure::database)?;
    drop(statement);
    for (kind, value) in candidates {
        let path = PathBuf::from(&value);
        let valid = match kind.as_str() {
            "workspace" => load_workspace(&path).is_ok(),
            "package" => inspect_package(&path, &VerificationLimits::default()).is_ok(),
            _ => false,
        };
        if !valid {
            connection
                .execute(
                    "DELETE FROM recent_items WHERE kind = ?1 AND path = ?2",
                    params![kind, value],
                )
                .map_err(BridgeFailure::database)?;
        }
    }
    read_state(&connection)
}

fn open(database: &Path) -> Result<Connection, BridgeFailure> {
    if let Some(parent) = database.parent() {
        fs::create_dir_all(parent).map_err(|error| {
            BridgeFailure::new(
                "APP_STATE_DIRECTORY_FAILED",
                error.to_string(),
                Some(parent.to_path_buf()),
            )
        })?;
    }
    let connection = Connection::open(database).map_err(BridgeFailure::database)?;
    connection
        .busy_timeout(Duration::from_secs(5))
        .map_err(BridgeFailure::database)?;
    connection
        .execute_batch("PRAGMA foreign_keys = ON; PRAGMA journal_mode = WAL;")
        .map_err(BridgeFailure::database)?;
    Ok(connection)
}

fn migrate(connection: &Connection) -> Result<(), BridgeFailure> {
    let version = connection
        .pragma_query_value(None, "user_version", |row| row.get::<_, i64>(0))
        .map_err(BridgeFailure::database)?;
    if version > SCHEMA_VERSION {
        return Err(BridgeFailure::input(
            "APP_STATE_VERSION_UNSUPPORTED",
            "Application state database is newer than this workbench.",
        ));
    }
    if version == 0 {
        connection
            .execute_batch(
                "BEGIN;
                 CREATE TABLE preferences(
                   key TEXT PRIMARY KEY NOT NULL,
                   value TEXT NOT NULL
                 );
                 CREATE TABLE recent_items(
                   kind TEXT NOT NULL CHECK(kind IN ('workspace', 'package')),
                   path TEXT NOT NULL,
                   last_opened_at TEXT NOT NULL,
                   PRIMARY KEY(kind, path)
                 );
                 PRAGMA user_version = 1;
                 COMMIT;",
            )
            .map_err(BridgeFailure::database)?;
    }
    Ok(())
}

fn read_state(connection: &Connection) -> Result<WorkbenchState, BridgeFailure> {
    let mut preferences = BTreeMap::new();
    let mut preference_statement = connection
        .prepare("SELECT key, value FROM preferences ORDER BY key")
        .map_err(BridgeFailure::database)?;
    for row in preference_statement
        .query_map([], |row| {
            Ok((row.get::<_, String>(0)?, row.get::<_, String>(1)?))
        })
        .map_err(BridgeFailure::database)?
    {
        let (key, value) = row.map_err(BridgeFailure::database)?;
        preferences.insert(key, value);
    }
    Ok(WorkbenchState {
        schema_version: connection
            .pragma_query_value(None, "user_version", |row| row.get(0))
            .map_err(BridgeFailure::database)?,
        preferences,
        recent_workspaces: read_recents(connection, "workspace")?,
        recent_packages: read_recents(connection, "package")?,
    })
}

fn read_recents(connection: &Connection, kind: &str) -> Result<Vec<RecentItem>, BridgeFailure> {
    let mut statement = connection
        .prepare(
            "SELECT path, last_opened_at FROM recent_items \
             WHERE kind = ?1 ORDER BY last_opened_at DESC, rowid DESC LIMIT 20",
        )
        .map_err(BridgeFailure::database)?;
    statement
        .query_map(params![kind], |row| {
            Ok(RecentItem {
                path: row.get(0)?,
                last_opened_at: row.get(1)?,
            })
        })
        .map_err(BridgeFailure::database)?
        .collect::<Result<Vec<_>, _>>()
        .map_err(BridgeFailure::database)
}

fn validate_kind(kind: &str) -> Result<(), BridgeFailure> {
    if matches!(kind, "workspace" | "package") {
        Ok(())
    } else {
        Err(BridgeFailure::input(
            "RECENT_KIND_NOT_ALLOWED",
            "Recent item kind must be workspace or package.",
        ))
    }
}

#[cfg(test)]
pub fn contains_recent(database: &Path, kind: &str, path: &Path) -> Result<bool, BridgeFailure> {
    let connection = open(database)?;
    migrate(&connection)?;
    connection
        .query_row(
            "SELECT 1 FROM recent_items WHERE kind = ?1 AND path = ?2",
            params![kind, path.to_string_lossy()],
            |_| Ok(true),
        )
        .optional()
        .map(|value| value.unwrap_or(false))
        .map_err(BridgeFailure::database)
}

File: apps/desktop/src/main/app-state.ts
import fs from "node:fs";
import path from "node:path";
import { DatabaseSync } from "node:sqlite";

export interface StoredRecentItem {
  kind: "workspace" | "package";
  path: string;
  lastOpenedAt: string;
}

export interface StoredWorkbenchState {
  schemaVersion: number;
  preferences: Record<string, string>;
  recentWorkspaces: StoredRecentItem[];
  recentPackages: StoredRecentItem[];
}

const SCHEMA_VERSION = 1;

export class WorkbenchStateStore {
  #database: DatabaseSync;

  constructor(private readonly databasePath: string) {
    fs.mkdirSync(path.dirname(databasePath), { recursive: true });
    this.#database = this.#openWithRecovery();
  }

  read(): StoredWorkbenchState {
    const preferences: Record<string, string> = {};
    for (const raw of this.#database
      .prepare("SELECT key, value FROM preferences ORDER BY key")
      .all()) {
      const row = raw as { key: string; value: string };
      preferences[row.key] = row.value;
    }
    const recents = (kind: "workspace" | "package") =>
      this.#database
        .prepare(
          "SELECT kind, path, last_opened_at AS lastOpenedAt FROM recent_items " +
            "WHERE kind = ? ORDER BY last_opened_at DESC, rowid DESC LIMIT 20",
        )
        .all(kind) as StoredRecentItem[];
    return {
      schemaVersion: SCHEMA_VERSION,
      preferences,
      recentWorkspaces: recents("workspace"),
      recentPackages: recents("package"),
    };
  }

  setPreference(key: string, value: string): StoredWorkbenchState {
    this.#database
      .prepare(
        "INSERT INTO preferences(key, value) VALUES(?, ?) " +
          "ON CONFLICT(key) DO UPDATE SET value = excluded.value",
      )
      .run(key, value);
    return this.read();
  }

  remember(
    kind: "workspace" | "package",
    selectedPath: string,
  ): StoredWorkbenchState {
    this.#database
      .prepare(
        "INSERT INTO recent_items(kind, path, last_opened_at) VALUES(?, ?, ?) " +
          "ON CONFLICT(kind, path) DO UPDATE SET last_opened_at = excluded.last_opened_at",
      )
      .run(kind, selectedPath, new Date().toISOString());
    this.#database
      .prepare(
        "DELETE FROM recent_items WHERE rowid IN (" +
          "SELECT rowid FROM recent_items WHERE kind = ? " +
          "ORDER BY last_opened_at DESC, rowid DESC LIMIT -1 OFFSET 20)",
      )
      .run(kind);
    return this.read();
  }

  candidates(): StoredRecentItem[] {
    const state = this.read();
    return [...state.recentWorkspaces, ...state.recentPackages];
  }

  publishValidated(
    valid: Array<{ kind: "workspace" | "package"; path: string }>,
  ): StoredWorkbenchState {
    const keys = new Set(valid.map((item) => `${item.kind}\0${item.path}`));
    this.#database.exec("BEGIN IMMEDIATE");
    try {
      for (const item of this.candidates()) {
        if (!keys.has(`${item.kind}\0${item.path}`)) {
          this.#database
            .prepare("DELETE FROM recent_items WHERE kind = ? AND path = ?")
            .run(item.kind, item.path);
        }
      }
      this.#database.exec("COMMIT");
    } catch (error) {
      this.#database.exec("ROLLBACK");
      throw error;
    }
    return this.read();
  }

  close(): void {
    this.#database.close();
  }

  #openWithRecovery(): DatabaseSync {
    try {
      return this.#open();
    } catch (error) {
      const suffix = `.corrupt-${Date.now()}`;
      for (const candidate of [
        this.databasePath,
        `${this.databasePath}-wal`,
        `${this.databasePath}-shm`,
      ]) {
        if (fs.existsSync(candidate))
          fs.renameSync(candidate, `${candidate}${suffix}`);
      }
      try {
        return this.#open();
      } catch {
        throw error;
      }
    }
  }

  #open(): DatabaseSync {
    const database = new DatabaseSync(this.databasePath, {
      enableForeignKeyConstraints: true,
      enableDoubleQuotedStringLiterals: false,
    });
    try {
      database.exec("PRAGMA journal_mode = WAL; PRAGMA busy_timeout = 5000;");
      const rawVersion = database.prepare("PRAGMA user_version").get() as {
        user_version: number;
      };
      if (rawVersion.user_version > SCHEMA_VERSION) {
        throw new Error("APP_STATE_VERSION_UNSUPPORTED");
      }
      if (rawVersion.user_version === 0) {
        database.exec(
          "BEGIN;" +
            "CREATE TABLE preferences(key TEXT PRIMARY KEY NOT NULL, value TEXT NOT NULL);" +
            "CREATE TABLE recent_items(" +
            "kind TEXT NOT NULL CHECK(kind IN ('workspace', 'package'))," +
            "path TEXT NOT NULL, last_opened_at TEXT NOT NULL, PRIMARY KEY(kind, path));" +
            "PRAGMA user_version = 1; COMMIT;",
        );
      }
      return database;
    } catch (error) {
      database.close();
      throw error;
    }
  }
}
